cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -85,7 +86,13 @@ Result<std::unordered_set<std::string>> GetColumnNames(
 
   RETURN_NOT_OK(
       parser.VisitLastRow([&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
-        util::string_view view{reinterpret_cast<const char*>(data), size};
+        // Skip BOM when reading column names (ARROW-14644)
+        ARROW_ASSIGN_OR_RAISE(auto data_no_bom, util::SkipUTF8BOM(data, size));
+        int32_t offset = data_no_bom - data;
+        DCHECK_GE(offset, 0);
+        size = size - offset;
```

Comment:
```suggestion
        ptrdiff_t offset = data_no_bom - data;
        DCHECK_GE(offset, 0);
        size = size - static_cast<uint32_t>(offset);
```
The Windows CI job is failing.

I think this will make windows happy but I never know and I don't have this in my IDE at the moment so I may be way off base here.

Reply:
Thanks, I wasn't sure what the best way to explicitly cast was. Those changes work locally on my Mac; hopefully will work in Windows builds.